### PR TITLE
Add unsigned operations to Int32, Int64, Nativeint

### PR DESCRIPTION
**Pull request by nojb**
I missed having unsigned operations in the standard integer modules lately, so following discussion of #1201, I propose here the addition of `compare_unsigned`, `div_unsigned`, and `rem_unsigned` to each of `Int32`, `Int64`, and `Nativeint`.

I did not focus on performance for a first version, so the implementation is in OCaml in terms of the existing signed operations, following Hacker's Delight, 2nd ed, section 9.3.

If this PR is considered for inclusion we should decide if we want to go one (or two) steps further and implement the operations with C primitives (easy) or even with special processor instructions (unlikely to be worth the trouble).

One more operation that may be useful is `of_int32_unsigned` (extension by zero).

### CODE REVIEW

**Comment by bluddy** [comment]
Right shift is also different for unsigned -- it doesn't check the sign bit but rather always places a 0 at the leftmost bit.

**Comment by alainfrisch** [comment]
> Right shift is also different for unsigned

`shift_right_logical` already exists.

File: stdlib/int32.ml
Comment (murmour) [inline stdlib/int32.ml:87]: Signed integer overflow is undefined, but you rely on a particular behaviour. Fishy-fishy.

A new C primitive would be a more robust solution.

Comment (xavierleroy) [reply stdlib/int32.ml:87]: > Signed integer overflow is undefined

What???  Overflows are perfectly defined in OCaml.  Undefined arithmetic is for lesser languages.

Comment (murmour) [reply stdlib/int32.ml:87]: As a frequent user of lesser languages, I'm delighted to hear these great news! The manual is totally silent on this important property of integer operations, so I assumed overflows to be undefined (as I've seen no definitions). Perhaps this should be clarified in the manual?

Comment (xavierleroy) [reply stdlib/int32.ml:87]: See http://caml.inria.fr/pub/docs/manual-ocaml/libref/Pervasives.html, section "Integer arithmetic":
> Integers are 31 bits wide (or 63 bits on 64-bit processors). All operations are taken modulo 2^31 (or 2^63). They do not fail on overflow.

Comment (murmour) [reply stdlib/int32.ml:87]: Oh. I didn't grep the standard library headers. My mistake.

Overflows seem to be perfectly defined, after all (and I'm still delighted). Thanks for correcting me.

**Comment by murmour** [comment]
I have carefully reviewed this PR and believe that it's a correct reference implementation.

**Comment by dbuenzli** [comment]
> One more operation that may be useful is of_int32_unsigned (extension by zero).

These kind of things would also be useful:

```ocaml
Int32.unsigned_to_int : int32 -> int option
Int64.unsigned_to_int : int64 -> int option
Nativeint.unsigned_to_int : nativeint -> int option
```

E.g. here's one I have in a project for the first function: 
```ocaml
let int31_of_uint32 =
  let max_int = Int32.of_int max_int in
  fun u ->
    if Int32.compare 0l u < 0 && Int32.compare u max_int <= 0
    then Some (Int32.to_int u)
    else None

let int63_of_uint32 =
  let move = Int64.to_int 0x1_0000_0000L in
  fun u ->
    let i = Int32.to_int u in
    Some (if i < 0 then i + move else i)

let int_of_uint32 = match Sys.word_size with
| 32 -> int31_of_uint32
| 64 -> int63_of_uint32
| _ -> assert false
```

**Comment by murmour** [comment]
@dbuenzli 
> E.g. here's one I have in a project for the first function:
> 
> ```ocaml
> let int31_of_uint32 =
>   let max_int = Int32.of_int max_int in
>   fun u ->
>     if Int32.compare 0l u < 0 && Int32.compare u max_int <= 0
>     then Some (Int32.to_int u)
>     else None
> ```

It returns `None` for zero. The first condition should probably be `Int32.compare u 0l >= 0`.

**Comment by dbuenzli** [comment]
> It returns None for zero. The first condition should probably be Int32.compare u 0l >= 0.

Thank for debugging ! One more reason that indicates it's best to have this things in a proper library.

**Comment by dbuenzli** [comment]
Btw. to motivate the functions: binary data formats often have fields with unsigned indices/offsets that index into binary blobs. Since the latter are read/mmaped as bigarrays or strings you need `int`s to index those, hence the conversion from sized unsigned ints to `int`.

File: stdlib/int32.mli
Comment (dbuenzli) [inline stdlib/int32.mli]: I would prefer the `Int32.unsigned_to_int` naming. While the result in `int` cannot be negative you are not treating the `int` as an unsigned `int`.

Comment (dbuenzli) [reply stdlib/int32.mli]: Ah sorry in fact you are but I don't think you should. As a said these operations are mainly useful for indexing arrays, this why they should return options if they exceed the positive range of `int`s.

Comment (nojb) [reply stdlib/int32.mli]: I don't think I am, did I make a mistake? In any case renamed `to_int_unsigned -> unsigned_to_int` as suggested.

**Comment by nojb** [comment]
Ping! Any opinions on the eventual integration of this kind of PR?

Incidentally, it was mentioned to me that this code would be useful for some upcoming work on Flambda (of course this by itself is hardly an argument for addition into the stdlib, but well...)

**Comment by hcarty** [comment]
Could/should the PR also include `Int32.unsigned_to_int64` and `Nativeint.unsigned_to_int64` to cover 32 bit platforms?

File: stdlib/int64.mli
Comment (dbuenzli) [inline stdlib/int64.mli]: Could you please document what happens if number can't be represented in as an unsigned number in the resulting `int`. Personally I would still argue that this should return an `int option`.

Comment (nojb) [reply stdlib/int64.mli]: Thanks for the keeping track of this! I changed the return type to `int option`, which I agree is better.

**Comment by hannesm** [comment]
As a regular user of unsigned fixed-width integers (mostly in protocol code), I'd welcome this PR to be merged.  I reviewed and tested the implementation.  This PR lacks tests for `rem_unsigned` (while `div_unsigned`, `compare_unsigned`, and `unsigned_to_int` are tested).

**Comment by xclerc** [comment]
I would say that the benefit of native support is significant.

On `amd64`, a trivial program that generates couples of
integers and performs unsigned divisions is 7 to 10% faster
with native support, while spending most of its time generating
the values.

A draft patch implementing the support of the various
back-ends is available [here](https://github.com/xclerc/ocaml/tree/unsigned_intops_native). The tests added by @nojb in
this PR pass for all architectures (using qemu). Inria's CI
`precheck` seems to fail for some Windows flavors; I need
to investigate these issues.

**Comment by nojb** [comment]
We need to converge on this PR. All opinions voiced so far seem to be for inclusion of these functions. So, if you are against, please speak up!

On the level of implementation, we need to decide if we want to include native backend support as per @xclerc's patch or stick to an OCaml implementation. Personally I feel we could merge the pure OCaml implementation first and include the more performant native backend support in a later PR.

**Comment by dbuenzli** [comment]
That's kind of a naming nit but since we have `XXX.unsigned_to_int` wouldn't it make more sense to settle on `XXX.unsigned_OP` for the other operations rather than `XXX.OP_unsigned` ?

File: stdlib/int32.mli
Comment (dbuenzli) [inline stdlib/int32.mli]: The `@since` indications need to to be updated.

**Comment by nojb** [comment]
Seems fine to me.

File: stdlib/int32.mli
Comment (dbuenzli) [inline stdlib/int32.mli]: Using `{!div}` here rather than `[div]` would provide the reader with a direct link. (The comment applies for more docstrings below).

Comment (nojb) [reply stdlib/int32.mli]: Thanks, fixed.

File: stdlib/int32.mli
Comment (nojb) [reply stdlib/int32.mli]: Thanks, fixed.

**Comment by alainfrisch** [comment]
Is anyone opposed in principle to this PR?  In particular, it would be good to hear to the opinion of those in favor of #1201.  Of course adding the "unsigned" functions now does not preclude in theory adding dedicated types in the future, but it would make the argument more difficult.

Considering that #1201 got a strong negative feedback from @xavierleroy and is thus unlikely to merged (soon), I think that the present PR is a good middle-ground.  Personally I think that dedicated types would be a good use of the type system to avoid mistakes, but the use cases for unsigned ints are narrow enough that loosing this safety is probably ok, considering the relative weights of the two approches.  So basically I would be fine with either #1201 or #1458, but we should do at least one of them.  With this PR, it becomes more straightforward to create external unsigned int libraries (and they will benefit from compiler support if/when added later).

Another question: should we also add similar functions to `Int`, for the sake of uniformity (even if they will probably be less useful)?

**Comment by mshinwell** [comment]
@alainfrisch I'm planning to handle this one -- I just haven't got to it yet.  I hope to have time soon.

**Comment by mshinwell** [comment]
@alainfrisch It would in fact be nice if we could get this into 4.08; do you have time to manage the inclusion?  I'm not sure I am going to have time before the freeze because I'm having to spend a lot of time on the gdb patches.  Here are some points from our internal discussions about this at JS:
- The support seems highly desirable.
- This PR appears to have been subjected to suitable review on the algorithmic details.
- It's probably best if this gets merged first, and then we will see about getting @xclerc 's native support merged in fairly short order after that.  We would like to see full native support.
- To reduce (but not eliminate) extra work if we later introduce proper types for the unsigned integers, it might be best to remove the "unsigned_" prefixes, and instead put the unsigned operations in sub-modules (which for the moment wouldn't declare "`type t`" but just be used for namespacing).

**Comment by alainfrisch** [comment]
>  I'm not sure I am going to have time before the freeze

(I'm not sure either -- and I don't know/want to know when the freeze will happen.  But I can try.)

> To reduce (but not eliminate) extra work if we later introduce proper types for the unsigned integers, it might be best to remove the "unsigned_" prefixes, and instead put the unsigned operations in sub-modules (which for the moment wouldn't declare "type t" but just be used for namespacing).

IIUC what you are suggesting, I think this would make future addition of unsigned types more difficult, actually.  If people start using e.g. `Int32.Unsigned.div`, passing them values of type `int32`, we won't be able to change the interface for that function later on without breaking code.

**Comment by mshinwell** [comment]
We're not going to be able to have zero breakage if proper types were introduced, clearly.  However if the names are correct now, all people need to do is to add the conversion functions, and nothing else.  If the names are incorrect, they will need to rename all of those too.

FWIW I think it's better style to have groups of functions like we have here in submodules anyway as a means of namespacing (in particular where the names of the operations then become "standard ones" such as one might see in some functor parameter's signature).

**Comment by alainfrisch** [comment]
> We're not going to be able to have zero breakage if proper types were introduced, clearly.

I don't see why.  We will just add (sub-)modules with the new types and related functions, and leave the existing *_unsigned around (they're relatively harmless, as their name is explicit enough), perhaps deprecated (or just with a comment to suggest using the unsigned types).

> FWIW I think it's better style to have groups of functions like we have here in submodules anyway as a means of namespacing

I'm not strongly opposed to putting new functions in `Int32.Unsigned`, but *if* we later add unsigned types, they should go somewhere else (e.g. `UInt32`).  I still see one risk of adding `Int32.Unsigned.div` : people might start opening that `Int32.Unsigned` in their code, which would have the ugly consequence of making the dynamic semantics depends on the `open`  statement (with no chance of catching problems with the type-checker).

**Comment by mshinwell** [comment]
Hmm, it seems strange to me that we would leave the existing functions lying around if we have proper types.  That aside though, the point about the module name is a good one; maybe it would be better to have separate compilation units if we had proper types.  I'm not really sure about that, but maybe we shouldn't commit either way yet, and just have the existing names.

**Comment by alainfrisch** [comment]
> Hmm, it seems strange to me that we would leave the existing functions lying around if we have proper types. 

As said, we could mark them as deprecated.  But I don't see a smooth migration plan if we want to reuse the names but change their type from version N to N+1.  Look at the difficulty to move from exception-raising functions to option-returning ones!

**Comment by bluddy** [comment]
I think it's a little strange that we all agree that having unsigned types makes sense except @xavierleroy. The benefits seem obvious, and it's really odd that we're hoping things will change later instead of making a strong case for why they're useful now. The whole point of the type system is to distinguish values that obey different rules, and unsigned values seems like a classic case for type distinction to me. The only argument I've heard for not having unsigned types is that it's less 'elegant', but in the real world, we have to use and interface with less elegant things. I think a strong caution against using these types in pure OCaml code that doesn't interface with the outside world is enough of a discouragement.

**Comment by pmetzger** [comment]
>  The only argument I've heard for not having unsigned types is that it's less 'elegant', but in the real world, we have to use and interface with less elegant things.

I have an argument for the "elegance" of unsigned ints that involves not calling them unsigned ints, but rather calling them "machine words". The operations for machine words are intended to be bit manipulation, while we think of integers as, well, integers, where the intended operations are supposed to be things like addition or multiplication.

If you're trying to count things or store a cardinal or what have you, you use an int. If you're trying to do cryptographic operations for which you want rotations or popcount or what have you, or you're trying to do bit manipulation on a GPIO port on an embedded system, and you're not really doing "arithmetic", you use a machine word.

Indeed, I'd argue for a system where doing bit manipulations on machine words doesn't signal errors on things like overflow, but where overflowing the integer representation for an integer _does_ signal an error — the former is probably not a bug in your program (or if it is, you asked for machine words!), while the latter probably _is_ a bug...

**Comment by XVilka** [comment]
There is more conventional name for this then - bit arrays or bitvectors.

**Comment by murmour** [comment]
How about naming the new functions `compare_as_unsigned`, `div_as_unsigned`, and `rem_as_unsigned`, to emphasize that they merely interpret their arguments as unsigned, without performing any type-level conversions?

If we ever get dedicated unsigned types, these "as"-functions will still stand on their own, providing a short-path to perform unsigned operations without explicit conversions.

**Comment by pmetzger** [comment]
@XVilka 
> There is more conventional name for this then - bit arrays or bitvectors.

Almost. When I think of bit vectors, I think of actual arrays of bits where the operations you're interested in are indexing, checking if the thing you are indexing is set or unset, and changing the bit at an index. This is very useful for densely packing booleans.

When I think of "machine word" operations, I'm really thinking of the combination of what you get from working in $\mathbb{Z}_{2^64}$ plus some non-traditional operations like rotate. This isn't really integers in the normal sense, but it isn't densely packed booleans either.

File: Changes
Comment (dbuenzli) [inline Changes]: You forgot to add the `unsigned_to_int` functions.

File: utils/targetint.mli
Comment (dbuenzli) [inline utils/targetint.mli]: This is not part of the stdlib but that `32-bit` is likely to be wrong.

**Review by dbuenzli — COMMENTED** [review]
Regarding #1201 vs this PR, I rather have this than nothing. Besides we at least know this one has @xavierleroy's [blessing](https://github.com/ocaml/ocaml/pull/1201#issuecomment-333941042) (unless he changed his mind).


Regarding submodule or not I personally don't see a strong need to namespace them in an `Unsigned` submodule and would rather be in favor of including the names as currently implemented in the PR.

File: testsuite/tests/basic/boxedints.ml
Comment (alainfrisch) [inline testsuite/tests/basic/boxedints.ml:95]: Reading this made me realize that the overflow behavior of `Int32.of_int` is not specified on 64-bit platforms.  Since the test depends on a truncation semantics (i.e. taking the result modulo 2^32), it might be worth documenting it.

**Review by alainfrisch — APPROVED** [review]
LGTM

File: Changes
Comment (nojb) [reply Changes]: Thanks, fixed.

File: utils/targetint.mli
Comment (nojb) [reply utils/targetint.mli]: Thanks, fixed.

File: testsuite/tests/basic/boxedints.ml
Comment (nojb) [reply testsuite/tests/basic/boxedints.ml:95]: Thanks, added.

**Comment by nojb** [comment]
Will merge this by end of day unless someone objects.

**Comment by gasche** [comment]
INRIA's CI reports a few build failures after this PR was merged ([example log](https://ci.inria.fr/ocaml/job/parallel-build/flambda=true,label=ocaml-linux-64/1077/consoleText)), and they look like failures caused by out-of-date `.depend` files:

```

../boot/ocamlrun ../ocamlopt -strict-sequence -absname -w +a-4-9-41-42-44-45-48 -g -warn-error A -bin-annot -nostdlib -safe-string -strict-formats -O3   \
           -p -c -o stdlib__unit.p.cmx unit.ml
File "/home/barsac/ci/builds/workspace/parallel-build/flambda/true/label/ocaml-linux-64/stdlib/_none_", line 1:
Error (warning 58): no cmx file was found in path for module Stdlib__sys, and its interface was not compiled with -opaque
```

I will do a `make depend` run, commit, and see whether it fixes it.

**Comment by gasche** [comment]
Doing a `make alldepend` did fix the issue, nothing to worry about here.

**Comment by nojb** [comment]
Thanks!

**Comment by yallop** [comment]
@alainfrisch:
> With this PR, it becomes more straightforward to create external unsigned int libraries

I don't see how that's the case, unless you ignore polymorphic comparison operations.  A library like [integers](https://github.com/ocamllabs/ocaml-integers) where unsigned integer types have proper comparison behaviour can't make use of this PR.

**Comment by alainfrisch** [comment]
> unless you ignore polymorphic comparison operations

Indeed, I did ignore them.  This needs to be documented of course, but I don't think this is necessarily a showstopper (personally I try to avoid polymorphic comparison -- not counting equality -- on anything except `int`).

**Comment by alainfrisch** [comment]
Note that the [Num library](https://github.com/ocaml/num/blob/master/src/num.ml) also exposes numeric types for which the polymorphic comparison does not correspond to the natural ordering (namely `Num`, `Ratio`, `Big_int`).  This was part of the standard distribution for a long time without people overly complaining about that specific mis-feature.

**Comment by yallop** [comment]
I'm not sure about that.  People did complain and @xavierleroy  recommended using `Zarith` in place of `Num`, in part so that polymorphic comparison would behave properly:

[Mantis 5375](https://caml.inria.fr/mantis/view.php?id=5375):
> > [edmcman] Since compare does not work with big ints, we must manually write our own comparison functions, which is very unsatisfying (and tedious).

> [xleroy] I can only recommend that you replace big_int by Zarith in your code [...] it also supports polymorphic comparisons perfectly. 

[Mantis 7341](https://caml.inria.fr/mantis/view.php?id=7341):
>> [goswin] The types defined by the num library can not be compared with Pervasives because the nat modules does not provide a compare function in the custom ops.

> [xleroy] As mentioned by doligez, the Zarith library provides big integers with correct polymorphic comparisons  [...] So, just use Zarith and don't expect Num to be fixed. 

The behaviour of polymorphic comparison for `Big_int` is rightly considered a bug, albeit one that's too inconvenient to fix.  Users will (rightly) consider broken polymorphic comparison a bug in unsigned integer implementations, too.

So, unfortunately, this PR only makes it easier to write buggy implementations of unsigned modules, and doesn't help with writing correct implementations.

**Comment by alainfrisch** [comment]
> The behaviour of polymorphic comparison for Big_int is rightly considered a bug,

I agree it's a problem, but it's not a bug.  If OCaml was a good language to implement an arbitrary precision library, that library would likely have the same problem.  And Zarith's `Q` module obviously has it!

As long as the language itself does not allow to use a custom comparison function for OCaml datatype themselves, one should rather manage users' expectation about the behavior of the polymorphic comparison function.

**Comment by yallop** [comment]
> I agree it's a problem, but it's not a bug.

Well, okay.  At the moment the `integers` library doesn't have this problem, and this PR doesn't make the implementation of `integers` any easier. However, the PR does make it more straightforward to create libraries with problems, which doesn't seem like a strong recommendation.

**Pull request merged**
